### PR TITLE
Support for companion objects

### PR DESCRIPTION
**Pull request by pikinier20**

### CODE REVIEW

File: src/main/scala/dotty/dokka/ScalaDocumentableToPageTranslator.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import org.jetbrains.dokka.base.signatures.SignatureProvider
+import org.jetbrains.dokka.base.transformers.pages.comments.CommentsToContentConverter
+import org.jetbrains.dokka.model.DModule
+import org.jetbrains.dokka.pages.ModulePageNode
+import org.jetbrains.dokka.transformers.documentation.DocumentableToPageTranslator
+import org.jetbrains.dokka.utilities.DokkaLogger
+
+class ScalaDocumentableToPageTranslator(
+    private val commentsToContentConverter: CommentsToContentConverter,
+    private val signatureProvider: SignatureProvider,
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/ScalaDocumentableToPageTranslator.scala]: `private val` should not be needed

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import org.jetbrains.dokka.model.properties._
+
+
+class ScalaPageCreator(
+    commentsToContentConverter: CommentsToContentConverter,
+    signatureProvider: SignatureProvider,
+    val logger: DokkaLogger
+) extends DefaultPageCreator(commentsToContentConverter, signatureProvider, logger) {
+    override def pageForClasslike(c: DClasslike): ClasslikePageNode = {
+        val res = super.pageForClasslike(c)
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/ScalaPageCreator.scala:20]:
```
super.pageForClasslike(c) match
  case c: DClass => 
     clazz.get(ClasslikeExtension).fold(c){ co =>
            res.modified(
                    res.getName,
                    res.getContent,
                    res.getDri,
                    res.getEmbeddedResources,
                    (res.getChildren.asScala ++ List(
                        ClasslikePageNode(
                            co.getName + "$",
                            contentForClasslike(co),
                            Set(co.getDri).asJava,
                            co,
                            (co.getClasslikes.asScala.map(pageForClasslike(_)) ++
                            co.getFunctions.asScala.map(pageForFunction(_))).asJava,
                            Nil.asJava
                        )
                    )).asJava
                )
     }
```

Comment (romanowski) [reply src/main/scala/dotty/dokka/ScalaPageCreator.scala:20]: Patternmatching is usually much more readable then `isInstanceOf` and `asInstanceOf`
`Option.fold` is nicer then `Option.map().getOrElse()`

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                        )
+                    )).asJava
+                )
+            ).getOrElse(res)
+        } else res
+    }
+
+    override def contentForClasslike(c: DClasslike): ContentGroup = {
+        val defContent = super.contentForClasslike(c)
+        if(c.isInstanceOf[DClass]){
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/ScalaPageCreator.scala]: Same as above (pattern match and fold)

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            val clazz = c.asInstanceOf[DClass]
+            clazz.get(ClasslikeExtension).companion.map(co => {
+                val addedContent = PageContentBuilder(commentsToContentConverter, signatureProvider, logger).contentFor(
+                    c,
+                    ContentKind.Main,
+                    Set().asJava,
+                    PropertyContainer.Companion.empty(),
+                    c.getSourceSets, builder => {
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/ScalaPageCreator.scala]: We can pull this lambda to inner method so it is more readble

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 5]
```diff
@@ -50,7 +50,18 @@ trait ClassLikeSupport:
 
     val typeDefs = classDef.body.collect { case td: TypeDef if !td.symbol.flags.is(Flags.Private) => td }
     val valDefs = classDef.body.collect { case td: ValDef if !isSyntheticField(td.symbol)  => td}
-    
+
+    val companion = if(kind != Kind.Object) {
+      if(!flags.is(Flags.Case)){
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: Case class also can have a custom companion objects

File: src/main/scala/dotty/dokka/tasty/TastyParser.scala [hunk 6]
```diff
@@ -14,6 +14,7 @@ import org.jetbrains.dokka.plugability.DokkaContext
 import dokka.java.api._
 import collection.JavaConverters._
 import org.jetbrains.dokka.model.properties.PropertyContainer
+import org.jetbrains.dokka.model.properties.{WithExtraProperties}
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/tasty/TastyParser.scala:17]: `{}` is not needed here

File: src/main/scala/dotty/dokka/tasty/TastyParser.scala [hunk 7]
```diff
@@ -38,7 +39,8 @@ case class TastyParser(reflect: Reflection, inspector: DokkaTastyInspector)
               docs += parsePackage(pck)
             case packageObject: ClassDef if(packageObject.symbol.name == "package$") =>
               docs += parsePackageObject(packageObject)
-            case clazz: ClassDef  =>
+            case clazz: ClassDef if(clazz.symbol.companionClass.exists && clazz.symbol.flags.is(Flags.Object)) =>
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/tasty/TastyParser.scala]: We can move that if to rhs of case below:
```
 case clazz: ClassDef  =>
   if clazz.symbol.companionClass.exists && clazz.symbol.flags.is(Flags.Object) then
     docs += parseClass(clazz)
case _ =>
```

File: src/main/scala/tests/genericMethods.scala [hunk 8]
```diff
@@ -9,8 +9,8 @@ class Types:
     def mixedAndAndOr(base: Int & String, nested: List[Int | Double] & Seq[String]): Unit 
     = ???
 
-    def literal(i: 1, d: 3.3, c: 'a'): 34 
-    = 34
+    def literal(i: 1, d: 3.3, c: Char): 34 
+    = 34 /* TODO: Deleted default value to repair searchbar until fix is merged */
```
Comment (romanowski) [inline src/main/scala/tests/genericMethods.scala]: This is not default value but type signature (so `c` is always `'a'`).

Please create a task here and link its number using #taskNumer

File: src/main/scala/tests/objectSignatures.scala [hunk 9]
```diff
@@ -2,6 +2,10 @@ package tests
 package objectSignatures
 
 class A[T]
+{
+    val a: String = "asd"
+    def method3() = "asd"
+}
 
 object A
```
Comment (romanowski) [inline src/main/scala/tests/objectSignatures.scala:10]: Can we add some membters here to tests if we render doc for it?

**Review by romanowski — CHANGES_REQUESTED** [review]
We need better testing around members in compation objects and make sure that we document companion objects for case classes 

Can we test case objects as well?

File: src/main/scala/dotty/dokka/ScalaDocumentableToPageTranslator.scala (hunk 1 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/ScalaDocumentableToPageTranslator.scala]: Changed to vals

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala (hunk 2 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/ScalaPageCreator.scala:20]: Changed if elses to patternmatching - It's really much more readable

File: src/main/scala/dotty/dokka/ScalaPageCreator.scala (hunk 4 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/ScalaPageCreator.scala]: Done

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 5 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: Repaired

File: src/main/scala/dotty/dokka/tasty/TastyParser.scala (hunk 7 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/TastyParser.scala]: Reversed condition and moved down

File: src/main/scala/tests/genericMethods.scala (hunk 8 above)
Comment (pikinier20) [reply src/main/scala/tests/genericMethods.scala]: Created

File: src/main/scala/tests/objectSignatures.scala (hunk 9 above)
Comment (pikinier20) [reply src/main/scala/tests/objectSignatures.scala:10]: Added some tests

File: dokkaJavaApi/src/main/kotlin/dokka.java.api.JavaPlugin.kt [hunk 10]
```diff
@@ -65,8 +66,37 @@ abstract class JavaDokkaPlugin : DokkaPlugin() {
...
     abstract fun createResourceInstaller(ctx: DokkaContext) : PageTransformer
     abstract fun createEmbeddedResourceAppender(ctx: DokkaContext) : PageTransformer
+    abstract fun createDocumentableToPageTranslator(
+        commentsToContentConverter: CommentsToContentConverter,
+        signatureProvider: SignatureProvider,
+        logger: DokkaLogger
+    ) : DocumentableToPageTranslator
+}
+
+// TODO we probably does not need that
```
Comment (romanowski) [inline dokkaJavaApi/src/main/kotlin/dokka.java.api.JavaPlugin.kt:90]: Why we want to introduce it back?

File: src/main/scala/dotty/dokka/tasty/TastyParser.scala [hunk 11]
```diff
@@ -29,25 +30,26 @@ case class TastyParser(reflect: Reflection, inspector: DokkaTastyInspector)
...
-            super.traverseTree(tree)
-          
+        override def traverseTree(tree: Tree)(using ctx: Context): Unit = 
+          tree match {
+            case pck: PackageClause => 
+              docs += parsePackage(pck)
+              super.traverseTree(tree)
+            case packageObject: ClassDef if(packageObject.symbol.name == "package$") =>
+              docs += parsePackageObject(packageObject)
+            case clazz: ClassDef if(!clazz.symbol.isCompanionObject()) =>
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/tasty/TastyParser.scala:43]: `()` are not needed in ifs in cases

**Pull request merged**